### PR TITLE
Add bipedal-locomotion-framework

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
This is the first test for having bipedal-locomotion-framework in the superbuild.
I'm pretty sure I missed something 

cc @traversaro @S-Dafarra

### CODE REVIEW

**Comment by traversaro** [comment]
Thanks! At a first glance it seem nice, let's see what is missing. For adding a new components the documentation as outline by the steps in https://github.com/robotology/robotology-superbuild/blob/master/doc/developers-faqs.md#how-to-add-a-new-profile, but let's first see if CI passes, and then we can go forward.

File: cmake/Buildbipedal-locomotion-framework.cmake
Comment (traversaro) [inline cmake/Buildbipedal-locomotion-framework.cmake:32]: Could it make sense to pass the options on enabling/disabling the dependencies explicitly?

File: cmake/Buildbipedal-locomotion-framework.cmake
Comment (traversaro) [inline cmake/Buildbipedal-locomotion-framework.cmake:7]: Probably in the CMake world we can stick to the CMake package name for consistency, so:
```suggestion
set(BipedalLocomotionFramework_DEPENDS "")
```

File: CMakeLists.txt
Comment (traversaro) [inline CMakeLists.txt:165]: The name here should match the one of the `Build<package>.cmake`  and the one passed to `ycm_ep_helper` . Even if not strictly necessary, it is a good idea to just use the cmake package name (so in this case `BipedalLocomotionFramework`), because `find_or_build_package` mirrors `find_package`, so if we have: 
~~~
find_package(BipedalLocomotionFramework)
~~~
we should also have: 
~~~
find_or_build_package(BipedalLocomotionFramework)
~~~

**Comment by traversaro** [comment]
The casadi error on Debian Buster are one of those that make you feel alive: 
~~~
2020-11-22T14:35:49.3238803Z CMake Deprecation Warning at CMakeLists.txt:31 (cmake_policy):
2020-11-22T14:35:49.3239537Z   The OLD behavior for policy CMP0011 will be removed from a future version
2020-11-22T14:35:49.3240053Z   of CMake.
2020-11-22T14:35:49.3240274Z 
2020-11-22T14:35:49.3240935Z   The cmake-policies(7) manual explains that the OLD behaviors of all
2020-11-22T14:35:49.3241680Z   policies are deprecated and that a policy should be set to OLD only under
2020-11-22T14:35:49.3242608Z   specific short-term circumstances.  Projects should be ported to the NEW
2020-11-22T14:35:49.3243301Z   behavior and not rely on setting a policy to OLD.
2020-11-22T14:35:49.3243625Z 
2020-11-22T14:35:49.3243818Z 
2020-11-22T14:35:49.3244321Z CMake Deprecation Warning at CMakeLists.txt:35 (cmake_policy):
2020-11-22T14:35:49.3245034Z   The OLD behavior for policy CMP0005 will be removed from a future version
2020-11-22T14:35:49.3245541Z   of CMake.
2020-11-22T14:35:49.3245767Z 
2020-11-22T14:35:49.3246437Z   The cmake-policies(7) manual explains that the OLD behaviors of all
2020-11-22T14:35:49.3247173Z   policies are deprecated and that a policy should be set to OLD only under
2020-11-22T14:35:49.3248110Z   specific short-term circumstances.  Projects should be ported to the NEW
2020-11-22T14:35:49.3248791Z   behavior and not rely on setting a policy to OLD.
2020-11-22T14:35:49.3249126Z 
2020-11-22T14:35:49.3249316Z 
2020-11-22T14:35:49.3249752Z CMake Error at CMakeLists.txt:99 (enable_language):
2020-11-22T14:35:49.3250410Z   The Ninja generator does not support Fortran using Ninja version
2020-11-22T14:35:49.3250813Z 
2020-11-22T14:35:49.3251081Z     1.8.2
2020-11-22T14:35:49.3251261Z 
2020-11-22T14:35:49.3251785Z   due to lack of required features.  Kitware has implemented the required
2020-11-22T14:35:49.3252538Z   features and they have been merged to upstream ninja for inclusion in Ninja
2020-11-22T14:35:49.3253426Z   1.10 and higher.  As of this version of CMake, Ninja 1.10 has not been
2020-11-22T14:35:49.3254170Z   released.  Meanwhile, Kitware maintains a branch of Ninja at:
2020-11-22T14:35:49.3254575Z 
2020-11-22T14:35:49.3255358Z     https://github.com/Kitware/ninja/tree/features-for-fortran#readme
2020-11-22T14:35:49.3255884Z 
2020-11-22T14:35:49.3256526Z   with the required features.  One may build ninja from that branch to get
2020-11-22T14:35:49.3257083Z   support for Fortran.
2020-11-22T14:35:49.3257347Z 
2020-11-22T14:35:49.3257521Z 
2020-11-22T14:35:49.3258115Z -- Configuring incomplete, errors occurred!
2020-11-22T14:35:49.3259290Z See also "/__w/robotology-superbuild/robotology-superbuild/build/external/casadi/CMakeFiles/CMakeOutput.log".

~~~

However, Debian Buster is one of those platforms for which we keep compatibility mostly for the Core software to run on the pc104 of old robots, probably we do not need for now to support `ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS` on it.

**Comment by traversaro** [comment]
> However, Debian Buster is one of those platforms for which we keep compatibility mostly for the Core software to run on the pc104 of old robots, probably we do not need for now to support ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS on it.

Anyhow, I just check and this is sweet: Debian Buster backports has indeed Ninja 1.10 (see https://packages.debian.org/buster-backports/ninja-build), so we just need to install it from the backports as we already to for CMake in https://packages.debian.org/buster-backports/ninja-build . However, if more problem emerge we can just avoid to test it on Debian Buster, even because now that we support conda we can easily compile our code on arbitrary old Linux distributions through it.

**Comment by GiulioRomualdi** [comment]
I think we need to change the idyntree tag

**Comment by traversaro** [comment]
> 
> 
> I think we need to change the idyntree tag

I will release iDynTree 2.0 soon in the next days, this should handle that. Instead we need to have a way to exclude some components from the release.

**Comment by GiulioRomualdi** [comment]
> Instead we need to have a way to exclude some components from the release.

I didn't get it

**Comment by traversaro** [comment]
> 
> 
> > Instead we need to have a way to exclude some components from the release.
> 
> I didn't get it

The release of the robotology-superbuild is coherent group of tags for all the subprojects, that are continuously tested together. As BFL still do not have a release for the 2020.11 release of the superbuild, I would like to exclude it from the release 2020.11, while including in the "rolling release" version of the superbuild that we typically used in the lab. However it was not a request for a PR, more a note for me.

**Comment by traversaro** [comment]
> 
> 
> I think we need to change the idyntree tag

Once this PR is approved: https://github.com/robotology/idyntree/pull/778 , I will release iDynTree 2.0.0 and then we will have a bfl-compatible iDynTree on both 2020.11 release, Stable and Unstable branches.

**Comment by GiulioRomualdi** [comment]
As discussed in  https://github.com/robotology/robotology-superbuild/issues/517#issuecomment-731210149 some dependencies can be installed with brew/vcpkg. We now should decide what to do for them. 

For instance, `qhull` can be installed with brew. Should the superbuild fail if `qhull` is missing in macOS while it will compile it in Linux?

**Comment by GiulioRomualdi** [comment]
On the `conda` profile `casadi` is failing because it's not able to find ipopt. However, ipopt is installed 
https://github.com/robotology/robotology-superbuild/blob/1a9cd37b4838bb2475c799f87a0f1b10754ff4a3/.github/workflows/ci.yml#L55

**Comment by traversaro** [comment]
I would disable for now the `ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS` in conda and Debian sid. Both are still highly experimental environments that are definitely not our first target.

**Comment by traversaro** [comment]
> 
> 
> As discussed in [#517 (comment)](https://github.com/robotology/robotology-superbuild/issues/517#issuecomment-731210149) some dependencies can be installed with brew/vcpkg. We now should decide what to do for them.
> 
> For instance, `qhull` can be installed with brew. Should the superbuild fail if `qhull` is missing in macOS while it will compile it in Linux?

At the moment I would for the simplest solution, i.e. compile everywhere, we can complexify it once the first PR is landed.

**Comment by GiulioRomualdi** [comment]
- on `Tags:Release202011@ubuntu-18.04@Release` is failing with
  ```
  BipedalLocomotion::Planners::TimeVaryingDCMPlanner::Impl::computeConstraintElementsECMP(const 
  BipedalLocomotion::Planners::ContactPhase&, casadi::DM&, casadi::DM&)’:
  2020-11-24T18:11:03.4963635Z /home/runner/work/robotology-superbuild/robotology-superbuild/robotology/BipedalLocomotionFramework/src/Planners/src/TimeVaryingDCMPlanner.cpp:312:53: error: unused variable ‘contactName’ [-Werror=unused-variable]
  2020-11-24T18:11:03.4965142Z          for (const auto& [contactName, activeContact] : contactPhase.activeContacts)
  2020-11-24T18:11:03.4971726Z                                                      ^
  2020-11-24T18:11:07.1228673Z cc1plus: some warnings being treated as errors
  2020-11-24T18:11:07.1376820Z src/Planners/CMakeFiles/Planners.dir/build.make:105: recipe for target 'src/Planners/CMakeFiles/Planners.dir/src/TimeVaryingDCMPlanner.cpp.o' failed
  2020-11-24T18:11:07.1378922Z CMakeFiles/Makefile2:980: recipe for target 'src/Planners/CMakeFiles/Planners.dir/all' failed
  2020-11-24T18:11:07.1383038Z make[5]: *** [src/Planners/CMakeFiles/Planners.dir/src/TimeVaryingDCMPlanner.cpp.o] Error 1
  2020-11-24T18:11:07.1384252Z Makefile:146: recipe for target 'all' failed
  2020-11-24T18:11:07.1385871Z make[4]: *** [src/Planners/CMakeFiles/Planners.dir/all] Error 2
  2020-11-24T18:11:07.1389338Z CMakeFiles/BipedalLocomotionFramework.dir/build.make:134: recipe for target 'robotology/BipedalLocomotionFramework/CMakeFiles/YCMStamp/BipedalLocomotionFramework-build' failed
  2020-11-24T18:11:07.1390862Z make[3]: *** [all] Error 2
  2020-11-24T18:11:07.1392036Z make[2]: *** [robotology/BipedalLocomotionFramework/CMakeFiles/YCMStamp/BipedalLocomotionFramework-build] Error 2
  2020-11-24T18:11:07.1393701Z make[1]: *** [CMakeFiles/BipedalLocomotionFramework.dir/all] Error 2
  2020-11-24T18:11:07.1395211Z CMakeFiles/Makefile2:547: recipe for target 'CMakeFiles/BipedalLocomotionFramework.dir/all' failed
  2020-11-24T18:11:07.1396710Z Makefile:111: recipe for target 'all' failed
  2020-11-24T18:11:07.1397361Z make: *** [all] Error 2
  2020-11-24T18:11:07.1416062Z ##[error]Process completed with exit code 2.
  ```
  This happens because the master version of `manif` is used. (I think something is missing in the release `yml` file). This problem has been solved in https://github.com/artivis/manif/pull/157 (`manif devel`)

**Comment by traversaro** [comment]
> 
> 
> I would disable for now the `ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS` in conda and Debian sid. Both are still highly experimental environments that are definitely not our first target.

As the situation is quickly growing complex for the relative compatibility of options and platform support, I am preparing a matrix to clarify which profiles are a supported on which platform, I am adding it after https://github.com/robotology/robotology-superbuild/pull/535 lands.

**Comment by traversaro** [comment]
> 
> 
>     * on `Tags:Release202011@ubuntu-18.04@Release` is failing with
>       ```
>       BipedalLocomotion::Planners::TimeVaryingDCMPlanner::Impl::computeConstraintElementsECMP(const 
>       BipedalLocomotion::Planners::ContactPhase&, casadi::DM&, casadi::DM&)’:
>       2020-11-24T18:11:03.4963635Z /home/runner/work/robotology-superbuild/robotology-superbuild/robotology/BipedalLocomotionFramework/src/Planners/src/TimeVaryingDCMPlanner.cpp:312:53: error: unused variable ‘contactName’ [-Werror=unused-variable]
>       2020-11-24T18:11:03.4965142Z          for (const auto& [contactName, activeContact] : contactPhase.activeContacts)
>       2020-11-24T18:11:03.4971726Z                                                      ^
>       2020-11-24T18:11:07.1228673Z cc1plus: some warnings being treated as errors
>       2020-11-24T18:11:07.1376820Z src/Planners/CMakeFiles/Planners.dir/build.make:105: recipe for target 'src/Planners/CMakeFiles/Planners.dir/src/TimeVaryingDCMPlanner.cpp.o' failed
>       2020-11-24T18:11:07.1378922Z CMakeFiles/Makefile2:980: recipe for target 'src/Planners/CMakeFiles/Planners.dir/all' failed
>       2020-11-24T18:11:07.1383038Z make[5]: *** [src/Planners/CMakeFiles/Planners.dir/src/TimeVaryingDCMPlanner.cpp.o] Error 1
>       2020-11-24T18:11:07.1384252Z Makefile:146: recipe for target 'all' failed
>       2020-11-24T18:11:07.1385871Z make[4]: *** [src/Planners/CMakeFiles/Planners.dir/all] Error 2
>       2020-11-24T18:11:07.1389338Z CMakeFiles/BipedalLocomotionFramework.dir/build.make:134: recipe for target 'robotology/BipedalLocomotionFramework/CMakeFiles/YCMStamp/BipedalLocomotionFramework-build' failed
>       2020-11-24T18:11:07.1390862Z make[3]: *** [all] Error 2
>       2020-11-24T18:11:07.1392036Z make[2]: *** [robotology/BipedalLocomotionFramework/CMakeFiles/YCMStamp/BipedalLocomotionFramework-build] Error 2
>       2020-11-24T18:11:07.1393701Z make[1]: *** [CMakeFiles/BipedalLocomotionFramework.dir/all] Error 2
>       2020-11-24T18:11:07.1395211Z CMakeFiles/Makefile2:547: recipe for target 'CMakeFiles/BipedalLocomotionFramework.dir/all' failed
>       2020-11-24T18:11:07.1396710Z Makefile:111: recipe for target 'all' failed
>       2020-11-24T18:11:07.1397361Z make: *** [all] Error 2
>       2020-11-24T18:11:07.1416062Z ##[error]Process completed with exit code 2.
>       ```
>       
>       
>       This happens because the master version of `manif` is used. (I think something is missing in the release `yml` file). This problem has been solved in [artivis/manif#157](https://github.com/artivis/manif/pull/157) (`manif devel`)

As in any case we are not going to add tags for bfl in 2020.11, I think we can skip that profile for now in release CIs, for example adding after all the configure steps: 
~~~
    - name: Disable profiles that do not support releases for now
      if: contains(matrix.project_tags, 'Release')
      run: | 
        cd build
        cmake -DROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS:BOOL=OFF .
~~~

**Comment by GiulioRomualdi** [comment]
> As in any case we are not going to add tags for bfl in 2020.11, 
ops I did it in https://github.com/robotology/robotology-superbuild/pull/526/commits/082ec057d5db93a940ab6c9de197d269eedfc2de
I remove it now

**Comment by traversaro** [comment]
> This happens because the master version of manif is used. (I think something is missing in the release yml file). This problem has been solved in artivis/manif#157 (manif devel) 



> 
> 
> > As in any case we are not going to add tags for bfl in 2020.11,
> 
> ops I did it in [082ec05](https://github.com/robotology/robotology-superbuild/commit/082ec057d5db93a940ab6c9de197d269eedfc2de)
> I remove it now

Ah cool, that's fine then, you can leave them, I did not think we wanted to do that for the moment, but if you did the effort we can leave them.

**Comment by GiulioRomualdi** [comment]
> Ah cool, that's fine then, you can leave them, I did not think we wanted to do that for the moment, but if you did the effort we can leave them.

let's see if they compile ;)

**Comment by GiulioRomualdi** [comment]
Something similar to this can be used to remove `ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS` from debian right?
> 
> ```
>     - name: Disable profiles that do not support releases for now
>       if: contains(matrix.project_tags, 'Release')
>       run: | 
>         cd build
>         cmake -DROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS:BOOL=OFF .
> ```

example
 ```
   - name: Disable profiles that do not support releases for now
      if: matrix.docker_image == 'debian:sid
      run: | 
        cd build
        cmake -DROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS:BOOL=OFF .
```

**Comment by traversaro** [comment]
> Something similar to this can be used to remove ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS from debian right?

Exactly!

**Comment by GiulioRomualdi** [comment]
Now in `conda` is failing because of https://github.com/dic-iit/bipedal-locomotion-framework/pull/152. https://github.com/dic-iit/bipedal-locomotion-framework/pull/153 should fix the problem

**Comment by GiulioRomualdi** [comment]
Unfortunately, I'm not able to find the error on windows. The log file is too heavy for my laptop

**Comment by traversaro** [comment]
> On macOS is failing because of ycm ?!

This is tracked in https://github.com/robotology/robotology-superbuild/pull/534 and hopefully fixed by https://github.com/robotology/ycm/pull/367, let's ignore it for now.

**Comment by GiulioRomualdi** [comment]
On `debian-buster`
```
CMake Error at CMakeLists.txt:99 (enable_language):
  The Ninja generator does not support Fortran using Ninja version

    1.8.2

  due to lack of required features.  Kitware has implemented the required
  features and they have been merged to upstream ninja for inclusion in Ninja
  1.10 and higher.  As of this version of CMake, Ninja 1.10 has not been
  released.  Meanwhile, Kitware maintains a branch of Ninja at:

    https://github.com/Kitware/ninja/tree/features-for-fortran#readme

  with the required features.  One may build ninja from that branch to get
  support for Fortran.
```

**Comment by traversaro** [comment]
> 
> 
> On `debian-buster` the following error is thrown while compiling casadi
> 
> ```
> CMake Error at CMakeLists.txt:99 (enable_language):
>   The Ninja generator does not support Fortran using Ninja version
> 
>     1.8.2
> 
>   due to lack of required features.  Kitware has implemented the required
>   features and they have been merged to upstream ninja for inclusion in Ninja
>   1.10 and higher.  As of this version of CMake, Ninja 1.10 has not been
>   released.  Meanwhile, Kitware maintains a branch of Ninja at:
> 
>     https://github.com/Kitware/ninja/tree/features-for-fortran#readme
> 
>   with the required features.  One may build ninja from that branch to get
>   support for Fortran.
> ```

Feel free to disable ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS  on buster

File: cmake/Buildbipedal-locomotion-framework.cmake
Comment (GiulioRomualdi) [reply cmake/Buildbipedal-locomotion-framework.cmake:32]: do you mean passing `FRAMEWORK_USE_manif` etc? https://github.com/dic-iit/bipedal-locomotion-framework/blob/master/cmake/BipedalLocomotionFrameworkFindDependencies.cmake#L179

Comment (GiulioRomualdi) [reply cmake/Buildbipedal-locomotion-framework.cmake:37]: https://github.com/robotology/robotology-superbuild/pull/526/commits/ab9935780819c036379b14c3e5d08cab9245bbd3

**Comment by GiulioRomualdi** [comment]
I also added `matio-cpp` that it's a dependency of blf. See https://github.com/robotology/robotology-superbuild/pull/526/commits/7e6f6041f5d3f65d4e37f4eda19aa1f6a29a7218 
cc @s-dafarra

**Comment by traversaro** [comment]
> 
> 
> I also added `matio-cpp` that it's a dependency of blf. See [7e6f604](https://github.com/robotology/robotology-superbuild/commit/7e6f6041f5d3f65d4e37f4eda19aa1f6a29a7218)
> cc @S-Dafarra

I think we need to add matio to the apt and brew dependencies in docs and CI, in vcpkg binary archive it is already available.

**Comment by GiulioRomualdi** [comment]
I realized that `ROBOTOLOGY_ENABLE_DYNAMICS_FULL_DEPS` was disabled in Windows CI.  https://github.com/robotology/robotology-superbuild/pull/526/commits/b62342babd2fc1d14a3a22f05aa023bfca2bcc32 fixes the problem 
Furthermore https://github.com/robotology/robotology-superbuild/pull/526/commits/a97dab519952e9e955ec95e86d68f07feb020a25 forces `casadi` to be installed following "normal" folder structure. This is required in windows. Please check [here](https://github.com/casadi/casadi/blob/fadc86444f3c7ab824dc3f2d91d4c0cfe7f9dad5/CMakeLists.txt#L317-L323)

**Comment by traversaro** [comment]
Casadi failures on Linux:
~~~
2020-11-26T11:01:14.0058780Z -- Install configuration: "Debug"
2020-11-26T11:01:14.0059403Z CMake Error at cmake_install.cmake:49 (file):
2020-11-26T11:01:14.0060296Z   file called with unknown argument "lib/cmake/casadi""".
~~~

**Comment by GiulioRomualdi** [comment]
@traversaro 
> Casadi failures on Linux:
> 
> ```
> 2020-11-26T11:01:14.0058780Z -- Install configuration: "Debug"
> 2020-11-26T11:01:14.0059403Z CMake Error at cmake_install.cmake:49 (file):
> 2020-11-26T11:01:14.0060296Z   file called with unknown argument "lib/cmake/casadi""".
> ```

It should be fixed in https://github.com/robotology/robotology-superbuild/pull/526/commits/91377a0866a4178213c9133705ad54d246e0f6e7

**Comment by GiulioRomualdi** [comment]
On windows is failing but I'm not able to find it. @traversaro could you help me?

**Review by traversaro — APPROVED** [review]

**Comment by GiulioRomualdi** [comment]
Everything is compiling!!!

**Comment by traversaro** [comment]
Thanks @GiulioRomualdi

**Pull request merged**
